plugins/synaptics-rmi/fu-synaptics-rmi-common.c

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2012-2014 Andrew Duggan
+ * Copyright (C) 2012-2019 Synaptics Inc.
+ * Copyright (C) 2019 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include <gio/gio.h>
+
+#include "fu-common.h"
+#include "fu-io-channel.h"
+#include "fu-synaptics-rmi-common.h"
+
+#include "fwupd-error.h"
+
+#define RMI_FUNCTION_QUERY_OFFSET		0
+#define RMI_FUNCTION_COMMAND_OFFSET		1
+#define RMI_FUNCTION_CONTROL_OFFSET		2
+#define RMI_FUNCTION_DATA_OFFSET		3
+#define RMI_FUNCTION_INTERRUPT_SOURCES_OFFSET	4
+#define RMI_FUNCTION_NUMBER			5
+
+#define RMI_FUNCTION_VERSION_MASK		0x60
+#define RMI_FUNCTION_INTERRUPT_SOURCES_MASK	0x7
+
+guint32
+fu_synaptics_rmi_generate_checksum (const guint8 *data, gsize len)
+{
+	guint32 lsw = 0xffff;
+	guint32 msw = 0xffff;
+	for (gsize i = 0; i < len / 2; i++) {
+		lsw += fu_common_read_uint16 (&data[i * 2], G_LITTLE_ENDIAN);
+		msw += lsw;
+		lsw = (lsw & 0xffff) + (lsw >> 16);
+		msw = (msw & 0xffff) + (msw >> 16);
+	}
+	return msw << 16 | lsw;
+}
+
+FuSynapticsRmiFunction *
+fu_synaptics_rmi_function_parse (GByteArray *buf, guint16 page_base, guint interrupt_count, GError **error)
+{
+	FuSynapticsRmiFunction *func = g_new0 (FuSynapticsRmiFunction, 1);
+	guint8 interrupt_offset;
+	const guint8 *data = buf->data;
+
+	/* not expected */
+	if (buf->len != RMI_DEVICE_PDT_ENTRY_SIZE) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "PDT entry buffer invalid size %u != %i",
+			     buf->len, RMI_DEVICE_PDT_ENTRY_SIZE);
+		return NULL;
```

Comment:
I think you have a memory leak here for func

Reply:
Gotcha, thanks.